### PR TITLE
[rfc] write explicit collectives in ShardingConfig

### PR DESCRIPTION
**Pull request by pianpwk**
proposal for alternative impl of https://github.com/pytorch/torchtitan/pull/3883, as override folder entry

- replaces dst_shardings placement in ShardingConfigs with per-axis redistribute, as overrideable `RedistributionSpec`. Deletes a chunk of spmd_types-specific logic, being more honest about what we support. Moves sharding.py files to this system

- allows the following WIP PR to implement the TP symm-mem AR as an override entry, adding a `@override(predicate=..)` field to only patch TP allreduces.

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #3936
* __->__ #3905

### CODE REVIEW

File: torchtitan/config/configurable.py [hunk 1]
```diff
@@ -128,6 +128,16 @@ def _traverse_child(val, prefix, parent, attr):
...
+                elif isinstance(val, dict):
+                    for key, item in val.items():
+                        item_fqn = f"{fqn}.{key}"
+                        if isinstance(item, config_cls):
+                            if recurse and hasattr(item, "traverse"):
+                                yield from _traverse_child(item, item_fqn, val, key)
+                            else:
+                                yield item_fqn, item, val, key
+                        elif hasattr(item, "traverse"):
+                            yield from _traverse_child(item, item_fqn, val, key)
```
Comment (pianpwk) [inline torchtitan/config/configurable.py]: because in_redist is dict[str, ShardingConfig], let overrides pattern match

File: torchtitan/config/override.py [hunk 2]
```diff
@@ -337,6 +342,8 @@ def apply_overrides(
         new_cfg = c.ov.factory(c.cfg)
         if isinstance(c.parent, list) and isinstance(c.attr, int):
             c.parent[c.attr] = new_cfg
+        elif isinstance(c.parent, dict):
+            c.parent[c.attr] = new_cfg
```
Comment (pianpwk) [inline torchtitan/config/override.py]: same as comment above

File: tests/unit_tests/test_parallel_dims.py [hunk 3]
```diff
@@ -380,34 +320,6 @@ def test_partition_spec_order_controls_state_shard(self):
...
-            result = spmd_redistribute_per_axis(
-                x,
-                mesh,
-                src.per_axis_spmd_types(),
-                dst.per_axis_spmd_types(),
-            )
-
-        self.assertEqual(comm_mode.get_total_counts(), 1)
-        self.assertTrue(torch.equal(result, torch.ones(2, 8, device=self.device_type)))
-
```
Comment (pianpwk) [inline tests/unit_tests/test_parallel_dims.py:410]: deleted these helpers

File: torchtitan/protocols/module.py [hunk 4]
```diff
@@ -417,43 +472,31 @@ def _maybe_wrap_with_local_region(
...
-                f"{type(self).__name__}: local_map is set but in_dst_shardings "
-                f"is missing entries for: {missing_in}"
-            )
-        in_named: list[SpmdLayout] = [in_dst[name] for name in pos_args]
 
         if parallel_dims.spmd_backend == "spmd_types":
-            return self._spmd_apply_local_map(fn, in_named, out_src)
-        return self._apply_local_map(fn, parallel_dims, in_named, out_src)
+            return self._spmd_apply_local_map(fn, out_src)
+        return self._apply_local_map(fn, parallel_dims, out_src)
```
Comment (pianpwk) [inline torchtitan/protocols/module.py:469]: so this no longer checks input placements for LocalMapConfig, but I think it's fine because

1) if there's no input redist, sharding config checks against in_src_placements already

2) if there is, we check in_src_placements, apply the collective, and then do local map, with nothing in between . I don't think much can go wrong here..

File: torchtitan/protocols/sharding.py [hunk 5]
```diff
@@ -62,8 +69,40 @@ def to_dict(self) -> dict:
...
+class RedistributionSpec(Configurable):
+    @dataclass(kw_only=True, slots=True)
+    class Config(Configurable.Config):
+        axis: MeshAxisName
+        src: spmd.PerMeshAxisSpmdType
+        dst: spmd.PerMeshAxisSpmdType
+        fwd_op_dtype: torch.dtype | None = None
+        fwd_out_dtype: torch.dtype | None = None
+        bwd_op_dtype: torch.dtype | None = None
+        bwd_out_dtype: torch.dtype | None = None
```
Comment (pianpwk) [inline torchtitan/protocols/sharding.py:88]: something to think about for comms: if we specify lower precision e.g. fp8, this just blindly casts, which seems a bit dumb to expose as a user config. it may be desirable to have basic scaling options (e.g. tensor/row/col/blockwise) if they're not hard to implement. but unsure if this should live in titan, or upstream, and what implementation to rely on.

cc @fegin

File: torchtitan/protocols/sharding.py [hunk 6]
```diff
@@ -90,10 +129,6 @@ class ShardingConfig:
             dataloader or FSDP-only path). Also declares the src side of
             the input redistribute pair.
             e.g. ``{"x": {TP: Shard(1)}}``.
-        in_dst_shardings: Desired input placements after redistribution,
```
Comment (tianyu-l) [inline torchtitan/protocols/sharding.py:93]: old docstrings are gone; new ones are missing

File: torchtitan/protocols/sharding.py [hunk 7]
```diff
@@ -62,8 +69,40 @@ def to_dict(self) -> dict:
         return {"repr": repr(self)}
 
 
+class RedistributionSpec(Configurable):
+    @dataclass(kw_only=True, slots=True)
+    class Config(Configurable.Config):
+        axis: MeshAxisName
```
Comment (tianyu-l) [inline torchtitan/protocols/sharding.py:82]: Two questions on the UX
1. If I want to redistribute on more than 1 axes, what do I do? E.g. TP all-gather followed by CP all-to-all. In previous `in_src -> in_dist` based configs, consecutive comms can happen on one axis after another.
2. If I'd like to perform fused comms, e.g. TP all-gather followed by CP all-gather, merged into one all-gather. Is that supported if I specify `axis="tp_cp"`, assuming I created such a flattened mesh ahead of time?

File: torchtitan/protocols/sharding.py [hunk 8]
```diff
@@ -62,8 +69,40 @@ def to_dict(self) -> dict:
         return {"repr": repr(self)}
 
 
+class RedistributionSpec(Configurable):
```
Comment (tianyu-l) [inline torchtitan/protocols/sharding.py]: Given current impl it probably should be
```suggestion
class PerAxisRedistribution(Function):
```
per https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/function.py

File: torchtitan/protocols/sharding.py [hunk 9]
```diff
@@ -62,8 +69,40 @@ def to_dict(self) -> dict:
...
+        self.config = config
+
+    def __call__(self, x: torch.Tensor) -> torch.Tensor:
+        axis = self.config.axis.value
+        if self.config.src == self.config.dst or spmd_mesh_size(axis) == 1:
+            return x
+        backward_options = {"op_dtype": self.config.bwd_op_dtype or x.dtype}
+        if self.config.bwd_out_dtype is not None:
+            backward_options["out_dtype"] = self.config.bwd_out_dtype
+        return spmd.redistribute(
```
Comment (tianyu-l) [inline torchtitan/protocols/sharding.py:100]: I could imagine user wanting to override this as well, e.g a custom all-reduce.

oh, I see in symm_mem_tp_all_reduce that you are introducing predicate

File: torchtitan/models/qwen3_5/sharding.py [hunk 10]
```diff
@@ -250,11 +254,18 @@ def _set_full_attention_sharding(
...
+    is_first_layer: bool,
 ) -> None:
     """TP sharding for Qwen35Attention (output gating + partial RoPE)."""
     attention_cfg.sharding_config = ShardingConfig(
         in_src_shardings={"x": attention_input_layout},
-        in_dst_shardings={"x": dense_activation_placement(tp=spmd.R)},
+        in_redist={
+            "x": RedistributionSpec.Config(
+                axis=MeshAxisName.TP,
+                src=spmd.R if is_first_layer else spmd.S(1),
```
Comment (tianyu-l) [inline torchtitan/models/qwen3_5/sharding.py:265]: The redundancy between `in_src_shardings` and `in_redist` src is on purpose? There's really no freedom in the latter.

oh I understood why we need this -- it's because you want to override the collective which you'd use predicate to understand

File: torchtitan/protocols/module.py [hunk 11]
```diff
@@ -544,18 +581,20 @@ def _redistribute_inputs(
         new_kwargs = dict(zip(pos_arg_names, args, strict=False))
         new_kwargs.update(kwargs)
 
-        in_dst_shardings = sharding_config.in_dst_shardings or {}
         in_src_shardings = sharding_config.in_src_shardings or {}
+        in_redist = sharding_config.in_redist or {}
+        in_redist_fns = self._in_redist_fns or {}
```
Comment (tianyu-l) [inline torchtitan/protocols/module.py]: why do you still need both, `in_redist_fns` is enough?

File: torchtitan/protocols/module.py [hunk 12]
```diff
@@ -18,23 +18,64 @@
...
+) -> Placement:
+    if axis_type == spmd.R or axis_type == spmd.I:
+        return Replicate()
+    if axis_type == spmd.P:
+        return Partial()
+    assert isinstance(axis_type, spmd.Shard)
+    return Shard(axis_type.dim)
+
+
+def _resolve_placements_after_redist(
```
Comment (tianyu-l) [inline torchtitan/protocols/module.py]: Is this util only for DTensor backends?

File: torchtitan/protocols/module.py [hunk 13]
```diff
@@ -18,23 +18,64 @@
...
 )
-from torchtitan.protocols.sharding import resolve_placements, ShardingConfig
+from torchtitan.protocols.sharding import (
+    RedistributionSpec,
+    resolve_placements,
+    ShardingConfig,
+)
+
+
+def _placement_from_spmd_type(
```
Comment (tianyu-l) [inline torchtitan/protocols/module.py]: Did we not have earlier or in spmd_types repo?

File: torchtitan/protocols/module.py [hunk 14]
```diff
@@ -280,10 +323,9 @@ def parallelize(self, parallel_dims: ParallelDims) -> None:
         if self._sharding_config is None:
             return
 
-        if parallel_dims.spmd_backend == "spmd_types":
-            spmd_validate_redistributions(self._sharding_config)
```
Comment (tianyu-l) [inline torchtitan/protocols/module.py:284]: I don't see why we can delete this check.

Let's say I have 
- in_src: `dp: S(0), tp: S(0)`
- in_redist: `dp axis, S(0) -> S(1)`

Is it not allowed before this PR but allowed now?

File: torchtitan/config/override.py [hunk 15]
```diff
@@ -101,6 +104,7 @@ def override(
     *,
     target: type[Configurable.Config],
     fqns: list[str] | None = None,
+    predicate: Callable[[Configurable.Config], bool] | None = None,
```
Comment (tianyu-l) [inline torchtitan/config/override.py]: without looking carefully, I have a question on conflict detection -- does conflict detection happen after filtering-by-predicate or before?
- e.g. override A wants to cover EP axis all-to-all, override B wants to cover TP axis all-gather. Would this be possible with current RP?

File: torchtitan/protocols/sharding.py (hunk 7 above)
Comment (fegin) [reply torchtitan/protocols/sharding.py:82]: > If I'd like to perform fused comms, e.g. TP all-gather followed by CP all-gather, merged into one all-gather. Is that supported if I specify axis="tp_cp", assuming I created such a flattened mesh ahead of time?

Do we have a concrete use case for fusing comms? I would imagine fusing comm with computation is more common.

**Review by fegin — COMMENTED** [review]
I like the idea of using sharding to replace the communication. But I'm worried that it is only useful for limited cases and will make the sharding configurations more complicated.  For TP use cases, it doesn't actually need `override`. The implementation can be applied unconditionally, TP or not TP. Using override is overkill. 

Another immediate use case is CP `selected_gather` case. However detecting whether `selected_gather` can be used or not require more information than `RedistributionSpec.__call__` has. So it is likely that I'll end up doing configuration programming to check the configuration to decide -- if CP is enabled, if the kernel is FlexAttention, the mask is `sliding window` or document causal. `RedistributionSpec` may still be useful.


I suggest that we should decouple `RedistributionSpec` from `override` usage. For communication, we should either put it into `distributed` folder, to maintain it. I'll do the same for `selected_gather`.

File: torchtitan/protocols/sharding.py (hunk 7 above)
Comment (pianpwk) [reply torchtitan/protocols/sharding.py:82]: > In previous in_src -> in_dist based configs, consecutive comms can happen on one axis after another.

technically this was only for DTensor, and titan doesn't have a case, but yes was going to add a comment on extending this

File: torchtitan/protocols/module.py (hunk 12 above)
Comment (pianpwk) [reply torchtitan/protocols/module.py]: yeah a bit awkward, it infers the dst placements then calls DTensor.redistribute on full placements, instead of doing per-axis

File: torchtitan/protocols/module.py (hunk 13 above)
Comment (pianpwk) [reply torchtitan/protocols/module.py]: my bad, we do, dropped this edit

File: torchtitan/protocols/module.py (hunk 14 above)
Comment (pianpwk) [reply torchtitan/protocols/module.py:284]: my bad, put partially back

**Comment by fegin** [comment]
After https://github.com/pytorch/pytorch/pull/189817, we can register python backend to do custom collectives. We still one flag from TorchTitan to enable the customization for allreduce but the override will not use what https://github.com/pytorch/torchtitan/pull/3883 does, but use the registered python backend. 

cc., @d4l3k @tianyu-l

File: torchtitan/protocols/sharding.py [hunk 16]
```diff
@@ -62,8 +69,47 @@ def to_dict(self) -> dict:
         return {"repr": repr(self)}
 
 
+class PerAxisRedistribution(Function):
```
Comment (tianyu-l) [inline torchtitan/protocols/sharding.py:72]: @fegin 
> I suggest that we should decouple `RedistributionSpec` from `override` usage. For communication, we should either put it into distributed folder, to maintain it. I'll do the same for selected_gather.

`override` is more for (1) demonstration runs (e.g. RL), or (2) usage outside the repo. I think what @pianpwk is doing on tp all-reduce override was aiming for (1). If the code is for us to maintain, I agree we should not use `override`.

> After https://github.com/pytorch/pytorch/pull/189817, we can register python backend to do custom collectives. We still one flag from TorchTitan to enable the customization for allreduce but the override will not use what https://github.com/pytorch/torchtitan/pull/3883 does, but use the registered python backend.

Is there a more concrete proposal for torchtitan / this PR?

My main questions still hold
- Does `torch.distributed` not care about such perf gap any more / admit we couldn't catch up, so that we are creating (first / another?) diverging point between pytorch and torchtitan today?
- Should we do such things in an annex repo? It sounds useful for beyond torchtitan.

Comment (fegin) [reply torchtitan/protocols/sharding.py:72]: > Does torch.distributed not care about such perf gap any more / admit we couldn't catch up, so that we are creating (first / another?) diverging point between pytorch and torchtitan today?

My answer to this question doesn't change. There is always a tradeoff between generality and performance. We are using one fixed symmetric buffer with fixed legth because our case is pretty limited. In reality all_reduce can be used in many cases, is the current design general enough for broader use cases. With such a small size (<4MB), the cost of generality is heavy. This can be shown when we scale the size, the fix overhead become negligible after certain sizes.

That being said, I agree this can be in the core PyTorch but it requires a more rigorous design to ensure it won't silently break all uses. I don't know if it worths though. cc., @d4l3k  

> Should we do such things in an annex repo? It sounds useful for beyond torchtitan.
For this particular one, I don't know if it worths a repo. Most if not all inference frameworks should already have their own implementations. Training is only benefit if the model is small but then you won't use TP for these training.

**Comment by tianyu-l** [comment]
I discussed more with @fegin and I proposed something "more frontier":
- how about we remove module boundary redistribution at all, but only specify module boundary sharding contract
- for TP / CP collectives, we explicitly call `spmd.redistribute` in the model code. That means we will have UlyssesAttention / AllGatherKVAttention, but it makes fused dist kernels much cleaner to support.
- Overall, we are seeing that frontier users would like to see things in the model code; and with spmd_types we can still hide all the if-else
